### PR TITLE
Use `teal.data::get_code` instead of `teal.data:::get_code_dependency`

### PR DESCRIPTION
**Pull request by m7pr**
After introduction of `...` in `teal.data::get_code()` during https://github.com/insightsengineering/teal.data/pull/290,
I reckon there is no point using non-exported `teal.data:::get_code_dependency` 
and we can just use exported `teal.data::get_code`.

Merge only when this is discussed and updated version of `teal.data` lands on CRAN https://github.com/insightsengineering/teal.data/issues/291

### CODE REVIEW

**Comment by m7pr** [comment]
Review only after this is merged https://github.com/insightsengineering/teal.data/pull/290

**Comment by averissimo** [comment]
Getting this in before {teal} is on CRAN would be nice to have 1 less `getFromNamespace()`

**Review by averissimo — CHANGES_REQUESTED** [review]
Blocking as requested by @m7pr 

Getting this in before {teal} is on CRAN would be nice to have 1 less `getFromNamespace()` in the package

**Comment by m7pr** [comment]
https://github.com/insightsengineering/teal.data/pull/290 is merged and `{teal.data}` 0.5.0 is released on https://github.com/insightsengineering/teal.data/pull/297/files

I guess we can merge this one. Restarting GHA builds

**Comment by m7pr** [comment]
3 tests are failing

**Comment by m7pr** [comment]
I think the quickest fix would be to create a `character` method for `teal.data::get_code` that would work directly on `character` code. So you could do as follows

- `teal.data::get_code(teal_data())  # using method for teal_data object`
- `teal.data::get_code(teal_data()@code) # using method for character`

**Comment by chlebowa** [comment]
> method for `teal.data::get_code` that would work directly on `character` code

Bad idea.

Let's discuss on a call.

**Comment by m7pr** [comment]
Or we can change the order of in `.datasets_to_data`

```r
  code <- c(
    get_rcode_str_install(),
    get_rcode_libraries(),
    get_datasets_code(datanames, datasets, hashes)
  )


  data <- do.call(
    teal.data::teal_data,
    args = c(data, code = list(code), join_keys = list(datasets$get_join_keys()[datanames]))
  )
````

So that `teal.data::teal_data()` is called and put into `get_datasets_code()`. Happy to discuss

File: R/utils.R [hunk 1]
```diff
@@ -68,7 +68,7 @@ teal_data_to_filtered_data <- function(x, datanames = teal_data_datanames(x)) {
     join_keys = teal.data::join_keys(x)
   )
   # Piggy-back entire pre-processing code so that filtering code can be appended later.
```
Comment (chlebowa) [inline R/utils.R:70]:
```suggestion
  # Piggy-back pre-processing code for datasets of interest so that filtering code can be appended later.
```

**Review by chlebowa — CHANGES_REQUESTED** [review]
Needs work after all.

May involve rewriting the `module_datasets` function or a completely different approach.

**Comment by m7pr** [comment]
Coolio. Just pasting the code we use to test the solution

```r
app <- init(
  data = teal_data(
    IRIS = iris, 
    MTCARS = mtcars, 
    code = "IRIS <- iris; x <-5; MTCARS <- mtcars"
  ),
  modules = modules(
    example_module(label = "MTCARS only", datanames = "MTCARS"),
    example_module(label = "IRIS only", datanames = "IRIS")
  ),
  filter = teal_slices(
    teal_slice("MTCARS", "mpg"),
    teal_slice("IRIS", "Species"),
    module_specific = TRUE,
    mapping = list(
      global_filters = c("MTCARS mpg", "IRIS Species")
    )
  )
)
if (interactive()) {
  shinyApp(app$ui, app$server)
}
```

**Comment by chlebowa** [comment]
It took me a while but:
I added a new file, `modules_datasets.R`. You can move the contents as you wish. The file contains
1. `modules_structure` - small functions that returns a nested list that reflects the structure of a `teal_modules`
2. `modules_datasets` - rather large function that returns list of `FilteredData` objects that reflects the shape of `teal_modules`
3. some testing, which can be easily adapted to the proper place in `module_teal.R` (uncomment)

File: R/modules_datasets.R [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  checkmate::assert_class(data, "teal_data")
+  checkmate::assert_multi_class(modules, c("teal_modules", "teal_module"))
+  checkmate::assert_class(filter, "modules_teal_slices")
+  checkmate::assert_r6(filtered_data_singleton, "FilteredData")
+
+  if (!isTRUE(attr(filter, "module_specific"))) {
+    slices <- shiny::isolate({
+      Filter(function(x) x$id %in% attr(filter, "mapping")$global_filters, filter)
+    })
+    singleton$set_filter_state(slices)
```
[the remark this answers is not in the document]
Comment (chlebowa) [reply R/modules_datasets.R]: Yes, sorry about that. Fixed.

**Comment by m7pr** [comment]
# Testing in the script

Hey, for `module_specific = TRUE` in your testing scenario I do get:

```r
$two
$two$example_two
[1] "mtcars cyl"

$two$three
$two$three$example_three
list()

$two$three$example_four
[1] "mtcars mpg" "mtcars cyl"



$example_one
[1] "iris Species"
```

Which looks proper. Of course for `module_specific = FALSE` I get

```r

$two
$two$example_two
[1] "mtcars cyl"

$two$three
$two$three$example_three
[1] "mtcars cyl"

$two$three$example_four
[1] "mtcars cyl"



$example_one
[1] "mtcars cyl"
```

which is expected. This part is great.


# Testing in the package

I tried to substitute `module_datasets()` with `modules_datasets()` https://github.com/insightsengineering/teal/pull/1115/files but got this error

```r
Warning in private$set_choices(slice$choices) :
  Choices adjusted (some values outside of variable range). Varname: mpg, dataname: MTCARS.
Warning: Error in ui_nested_tabs.teal_module: Assertion on 'datasets' failed: Must inherit from class 'FilteredData', but has class 'NULL'.
  3: runApp
  2: print.shiny.appobj
  1: <Anonymous>
```

while using 

```r
app <- init(
  data = teal_data(
    IRIS = iris, 
    MTCARS = mtcars, 
    code = "IRIS <- iris; x <-5; MTCARS <- mtcars"
  ),
  modules = modules(
    example_module(label = "all datasets", datanames = "all"),
    example_module(label = "MTCARS only", datanames = "MTCARS"),
    example_module(label = "IRIS only", datanames = "IRIS")
  ),
  filter = teal_slices(
    teal_slice("MTCARS", "mpg"),
    teal_slice("IRIS", "Species"),
    module_specific = TRUE,
    mapping = list(
      global_filters = c("MTCARS mpg", "IRIS Species")
    )
  )
)
if (interactive()) {
  shinyApp(app$ui, app$server)
}
```

**Comment by m7pr** [comment]
Hey @chlebowa I think we are good to merge.
We just need to 
a) remove commented code from `modules_datasets()`
b) decide where this functions lives
c) add tests for it?

**Comment by chlebowa** [comment]
Or do you want me to do it? :wink:

**Comment by m7pr** [comment]
@chlebowa sorry for the late reply. I can handle! No worries :)

**Comment by m7pr** [comment]
Hey @chlebowa I cleaned up your implementation and moved the function to utils.R. Added small roxygen documentation. Used your testing scenario in a test in testtthat folder. Lastly change the name of `filter` parameter, to `filters` so it plays with other parameters called `filters` in other functions that take `teal_slices` as an input.

**Comment by chlebowa** [comment]
> Lastly change the name of `filter` parameter, to `filters` so it plays with other parameters called `filters` in other functions that take `teal_slices` as an input.

Excellent :ok_hand:

File: R/utils.R [hunk 3]
```diff
@@ -220,6 +220,87 @@ check_filter_datanames <- function(filters, datanames) {
   }
 }
 
+
+#' Extract datasets for modules
```
Comment (chlebowa) [inline R/utils.R:224]:
```suggestion
#' Create filterable data for modules
#' 
#' Converts input data to a `FilteredData` object(s) to allow filtering before passing data to individual modules.
```

File: R/utils.R [hunk 4]
```diff
@@ -220,6 +220,87 @@ check_filter_datanames <- function(filters, datanames) {
...
+        )
+      } else {
+        modules$datanames
+      }
+    # 2. subset filters (global + dedicated)
+    slices <- shiny::isolate({
+      Filter(x = filters, f = function(x) {
+        x$dataname %in% datanames &&
+          (x$id %in% attr(filters, "mapping")$global_filters ||
+            x$id %in% unique(unlist(attr(filters, "mapping")[modules$label]))) # nolint
```
Comment (chlebowa) [inline R/utils.R:263]: Please specify linter exception.

File: R/utils.R [hunk 5]
```diff
@@ -220,6 +220,87 @@ check_filter_datanames <- function(filters, datanames) {
...
+#' If module specific, each leaf contains different instance, otherwise every leaf contains `filtered_data_singleton`.
+#' @keywords internal
+modules_datasets <- function(data, modules, filters, filtered_data_singleton = teal_data_to_filtered_data(data)) {
+  checkmate::assert_class(data, "teal_data")
+  checkmate::assert_multi_class(modules, c("teal_modules", "teal_module"))
+  checkmate::assert_class(filters, "modules_teal_slices")
+  checkmate::assert_r6(filtered_data_singleton, "FilteredData")
+
+  if (!isTRUE(attr(filters, "module_specific"))) {
+    slices <- shiny::isolate({
```
Comment (chlebowa) [inline R/utils.R:240]:
```suggestion
    # subset global filters
    slices <- shiny::isolate({
```

File: tests/testthat/test-utils.R [hunk 6]
```diff
@@ -64,6 +64,63 @@ test_that("teal_data_datanames returns datanames which are set by teal.data::dat
...
+      `example two` = "mtcars cyl",
+      three = list(
+        `example three` = list(),
+        `example four` = c("mtcars mpg", "mtcars cyl")
+      )
+    ),
+    `example one` = "iris Species"
+  )
+
+  testthat::expect_equal(modules_structure, expected_structure)
```
Comment (chlebowa) [inline tests/testthat/test-utils.R:121]: `expect_identical`?

**Review by chlebowa — APPROVED** [review]
Some minor comments, otherwise ready.

File: R/utils.R (hunk 4 above)
Comment (m7pr) [reply R/utils.R:263]: this is this ridiculous styler vs lintr infinite war : D

Comment (chlebowa) [reply R/utils.R]: I don't want an explanation, I want a specification. Which linter was turned off for this line?

Comment (m7pr) [reply R/utils.R]: got it. was trying to turn off `indentation_linter` https://github.com/insightsengineering/teal/pull/1118
put a PR for this

Comment (chlebowa) [reply R/utils.R]: pertect

**Pull request merged**